hashing/src/indexed_merkle_proof.rs

```diff
@@ -0,0 +1,557 @@
+
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields)]
+struct IndexedMerkleProofDeserializeValidator {
+    index: u64,
+    count: u64,
+    merkle_proof: Vec<Digest>,
+}
+
+impl TryFrom<IndexedMerkleProofDeserializeValidator> for IndexedMerkleProof {
+    type Error = error::MerkleConstructionError;
+    fn try_from(value: IndexedMerkleProofDeserializeValidator) -> Result<Self, Self::Error> {
+        let candidate = Self {
+            index: value.index,
+            count: value.count,
+            merkle_proof: value.merkle_proof,
+        };
+
+        if candidate.index > candidate.count
+            || candidate.merkle_proof.len() as u64 != candidate.compute_expected_proof_length()
+        {
+            return Err(error::MerkleConstructionError::IncorrectIndexedMerkleProof);
+        }
+        Ok(candidate)
+    }
+}
+
+#[derive(PartialEq, Debug, JsonSchema, Serialize, Deserialize)]
+#[serde(
+    deny_unknown_fields,
+    try_from = "IndexedMerkleProofDeserializeValidator"
+)]
+pub struct IndexedMerkleProof {
+    index: u64,
+    count: u64,
+    merkle_proof: Vec<Digest>,
+}
+
+impl ToBytes for IndexedMerkleProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, casper_types::bytesrepr::Error> {
+        let mut result = allocate_buffer(self)?;
+        result.append(&mut self.index.to_bytes()?);
+        result.append(&mut self.count.to_bytes()?);
+        result.append(&mut self.merkle_proof.to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.index.serialized_length()
+            + self.count.serialized_length()
+            + self.merkle_proof.serialized_length()
+    }
+}
+
+impl FromBytes for IndexedMerkleProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), casper_types::bytesrepr::Error> {
+        let ((index, count, merkle_proof), remainder) =
+            <(u64, u64, Vec<Digest>)>::from_bytes(bytes)?;
+
+        Ok((
+            IndexedMerkleProof::try_from(IndexedMerkleProofDeserializeValidator {
+                index,
+                count,
+                merkle_proof,
+            })
+            .map_err(|_| casper_types::bytesrepr::Error::Validation)?,
+            remainder,
+        ))
+    }
+}
+
+impl IndexedMerkleProof {
+    #[cfg(test)]
+    pub(crate) fn new<I>(
+        leaves: I,
+        index: u64,
+    ) -> Result<IndexedMerkleProof, error::MerkleConstructionError>
+    where
+        I: IntoIterator<Item = Digest>,
+    {
+        enum HashOrProof {
+            Hash(Digest),
+            Proof(Vec<Digest>),
+        }
+        use HashOrProof::{Hash, Proof};
+
+        let maybe_count_and_proof = leaves
+            .into_iter()
+            .enumerate()
+            .map(|(i, hash)| {
+                if i as u64 == index {
+                    (1u64, Proof(vec![hash]))
+                } else {
+                    (1u64, Hash(hash))
+                }
+            })
+            .tree_fold1(|(count_x, x), (count_y, y)| match (x, y) {
```

Comment:
We can use the `ExactSizeIterator` constraint here too.

Reply:
Implemented using `ExactSizeIterator`.